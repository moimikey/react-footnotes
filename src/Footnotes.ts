import React, { useEffect, ReactNode, useState, useRef } from "react";
import walker from "react-tree-walker";

export function Footnote({
  children,
  id,
}: {
  children: ReactNode;
  id: number;
}) {
  return React.createElement(React.Fragment, { key: id }, children);
}

Footnote.displayName = "Footnote";

export type FootnoteEntry = [number, ReactNode];
export type FootnotesChildren = (props: {
  Footnote: typeof Footnote;
  footnotes: Map<number, ReactNode>;
}) => ReactNode;

export function Footnotes(props: { children: FootnotesChildren }) {
  const initialized = useRef(false);
  const [footnotes, setFootnotes] = useState<Map<number, ReactNode>>(new Map());
  const newFootnotes = new Map<number, ReactNode>();

  const visit = (el: any) => {
    if (el.type && el.type.displayName === "Footnote") {
      const footnoteContent = Array.isArray(el.props.children)
        ? el.props.children[0]
        : el.props.children;
      newFootnotes.set(newFootnotes.size + 1, footnoteContent);
      setFootnotes(newFootnotes);
    }
  };

  useEffect(() => {
    if (!initialized.current) {
      initialized.current = true;
      walker(props.children({ Footnote, footnotes: newFootnotes }), visit);
    }
  }, []);

  return props.children({ Footnote, footnotes });
}

Footnotes.displayName = "Footnotes";

export default Footnotes;
