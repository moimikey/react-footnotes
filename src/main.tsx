"use strict";
import * as React from "react";
import * as ReactDOM from "react-dom/client";
import { Footnotes, FootnoteEntry } from ".";
import "./index.css";

const createFootnoteComponentWith = (FootnoteComponent: React.ElementType) => {
  return ({ children, id }: { children: React.ReactNode; id: number }) => {
    return (
      <FootnoteComponent>
        {children}
        <a href={`#footnote-${id}`}>
          <sup>{id}</sup>
        </a>
      </FootnoteComponent>
    );
  };
};

export function App() {
  return (
    <Footnotes>
      {({ Footnote, footnotes }) => {
        const Note = createFootnoteComponentWith(Footnote);
        return (
          <React.Fragment>
            <Note id={15}>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit.
            </Note>{" "}
            <Note id={22}>
              Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex ea commodo consequat.
            </Note>{" "}
            <Note id={35}>
              Duis aute irure dolor in reprehenderit in voluptate velit esse
              cillum dolore eu fugiat nulla pariatur.
            </Note>{" "}
            <Note id={42}>
              Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
              officia deserunt mollit anim id est laborum.
            </Note>{" "}
            <Note id={0}>Lorem ipsum</Note> <hr />
            References: should only show {footnotes.size}
            {Array.from(footnotes).map(([id, footnote]: FootnoteEntry) => (
              <div key={`footnote-${id}`} id={`footnote-${id}`}>
                <sup>{id}</sup> {footnote}
              </div>
            ))}
          </React.Fragment>
        );
      }}
    </Footnotes>
  );
}

let root = document.getElementById("root") as HTMLElement;

root.childNodes.length ||
  ReactDOM.createRoot(root).render(
    <React.StrictMode>
      <App />
    </React.StrictMode>
  );
