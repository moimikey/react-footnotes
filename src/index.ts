import { Footnotes } from "../lib/Footnotes";
import type { FootnoteEntry, FootnotesChildren } from "../lib/Footnotes";
export { Footnotes };
export type { FootnoteEntry, FootnotesChildren };
