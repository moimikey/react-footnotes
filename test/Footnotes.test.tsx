/**
 * @jest-environment jsdom
 */
// ^^^^^^^^^^^^^^^^^^^^^^^ DO NOT REMOVE
import React from "react";
import { createRoot } from "react-dom/client";
import { act } from "react-dom/test-utils";
import { Footnotes } from "../src";

let container: Element | null = null;

beforeEach(() => {
  if (!container) {
    container = document.createElement("div");
    document.body.appendChild(container);
  }
});

afterEach(() => {
  if (container) {
    document.body.removeChild(container);
    container = null;
  }
});

/**
 * NOTE: WRAP YOUR RENDERS IN <React.StrictMode> TO ENSURE IDEMPOTENCE!
 */

it("collects and provides footnotes correctly", async () => {
  let collectedFootnotes;
  const testFootnoteA =
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit.";
  const testFootnoteB =
    "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.";

  await act(() => {
    if (!container) {
      throw Error("Missing `container` element in DOM!");
    }
    createRoot(container).render(
      <React.StrictMode>
        <Footnotes>
          {({ Footnote, footnotes }) => {
            collectedFootnotes = footnotes;
            return (
              <React.Fragment>
                <Footnote id={1}>{testFootnoteA}</Footnote>
                <Footnote id={2}>{testFootnoteB}</Footnote>
              </React.Fragment>
            );
          }}
        </Footnotes>
      </React.StrictMode>
    );
  });
  expect(collectedFootnotes.size).toBe(2);
  expect(collectedFootnotes.get(1)).toBe(testFootnoteA);
  expect(collectedFootnotes.get(2)).toBe(testFootnoteB);
});

it("collects and provides footnotes correctly", async () => {
  const testFootnoteA =
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit.";
  const testFootnoteB =
    "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.";
  const expectedOutput = `Lorem ipsum dolor sit amet, consectetur adipiscing elit.Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.<div id=\"footnote-1\"><sup>1</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit.</div><div id=\"footnote-2\"><sup>2</sup> Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.</div>`;

  await act(() => {
    if (!container) {
      throw Error("Missing `container` element in DOM!");
    }
    createRoot(container).render(
      <React.StrictMode>
        <Footnotes>
          {({ Footnote, footnotes }) => {
            return (
              <React.Fragment>
                <Footnote id={1}>{testFootnoteA}</Footnote>
                <Footnote id={2}>{testFootnoteB}</Footnote>
                {Array.from(footnotes).map(([id, footnote]) => (
                  <div key={`footnote-${id}`} id={`footnote-${id}`}>
                    <sup>{id}</sup> {footnote}
                  </div>
                ))}
              </React.Fragment>
            );
          }}
        </Footnotes>
      </React.StrictMode>
    );
  });
  expect(container?.innerHTML).toBe(expectedOutput);
});
