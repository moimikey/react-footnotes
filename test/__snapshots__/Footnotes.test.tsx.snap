// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render 1`] = `
[
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
  "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
]
`;
